### PR TITLE
Add UI

### PR DESCRIPTION
**Pull request by VladislavSokov**
- [x] add icons to the buttons
- [x] move logic from controller to a service
- [x] refactor with "each db connection"
- [x] intro new env var for "disabled"

### CODE REVIEW

**Comment by VladislavSokov** [comment]
Screenshot from 2024-06-26 19-07-54
Screenshot from 2024-06-26 19-07-44

File: app/views/actual_db_schema/migrations/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>Phantom Migrations</title>
+    <%= stylesheet_link_tag 'actual_db_schema/styles', media: 'all' %>
```
Comment (ka8725) [inline app/views/actual_db_schema/migrations/index.html.erb:5]: Let's inline styles too. `stylesheet_link_tag` depends on configuration that can be broken in different Rails versions. That can make maintenance costs more expensive.

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+ActualDbSchema::Engine.routes.draw do
+  root to: "migrations#index"
+  resources :migrations, only: %i[index show] do
```
Comment (ka8725) [inline config/routes.rb:5]: Can we make the route as `/rails/migrations`?

File: lib/actual_db_schema/engine.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module ActualDbSchema
+  # It isolates the namespace to avoid conflicts with the main application.
+  class Engine < ::Rails::Engine
+    isolate_namespace ActualDbSchema
+
+    initializer "actual_db_schema.append_routes", after: "add_routing_paths" do |app|
+      app.routes.append do
+        mount ActualDbSchema::Engine => "/actual_db_schema"
```
Comment (ka8725) [inline lib/actual_db_schema/engine.rb:10]:
```suggestion
        mount ActualDbSchema::Engine => "/rails"
```

Can we do this? Won't it conflict with rails mailers previews?

File: app/views/actual_db_schema/migrations/index.html.erb (hunk 1 above)
Comment (ka8725) [reply app/views/actual_db_schema/migrations/index.html.erb:5]: I see it's used in the other file. So it makes sense to reuse styles like that. Please ignore the comment above.

File: config/routes.rb (hunk 2 above)
Comment (VladislavSokov) [reply config/routes.rb:5]: I'll try

File: lib/actual_db_schema/engine.rb [hunk 4]
```diff
@@ -5,14 +5,16 @@ module ActualDbSchema
...
+      if Rails.env.development? || Rails.env.test?
+        app.routes.append do
+          mount ActualDbSchema::Engine => "/rails"
+        end
       end
     end
 
     initializer "actual_db_schema.assets.precompile" do |app|
-      app.config.assets.precompile += %w[actual_db_schema/styles.css]
+      app.config.assets.precompile += %w[actual_db_schema/styles.css] if Rails.env.development?
```
Comment (ka8725) [inline lib/actual_db_schema/engine.rb:17]: Why not to do that also in test env?

File: lib/actual_db_schema/engine.rb [hunk 5]
```diff
@@ -5,14 +5,16 @@ module ActualDbSchema
   class Engine < ::Rails::Engine
     isolate_namespace ActualDbSchema
 
-    initializer "actual_db_schema.append_routes", after: "add_routing_paths" do |app|
-      app.routes.append do
-        mount ActualDbSchema::Engine => "/rails"
+    initializer "actual_db_schema.append_routes", after: :add_routing_paths do |app|
+      if Rails.env.development? || Rails.env.test?
```
Comment (ka8725) [inline lib/actual_db_schema/engine.rb:9]: WDYT if we introduce a separate config similar to "enabled" that will be responsible for loading the engine? By default, it's true when it's dev or test env. Also, what if preventing the loading engine at a higher level just not to have this "if" repeated in each block?

File: app/controllers/actual_db_schema/migrations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+      build_migration_hash(status, migration.version.to_s, migration)
+    end
+
+    def prepare_context
+      context = fetch_migration_context
+      context.extend(ActualDbSchema::Patches::MigrationContext)
+      context
+    end
+
+    def build_migration_hash(status, version, migration)
```
Comment (ka8725) [inline app/controllers/actual_db_schema/migrations_controller.rb:78]: Use struct, but not hashes for this please

File: app/controllers/actual_db_schema/migrations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      else
+        ActiveRecord::Base.connection_config
+      end
+    end
+
+    def branch_for(version)
+      metadata.fetch(version.to_s, {})[:branch] || "unknown"
+    end
+
+    def metadata
```
Comment (ka8725) [inline app/controllers/actual_db_schema/migrations_controller.rb:111]: No need to cache that?

File: app/controllers/actual_db_schema/migrations_controller.rb [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      database = params[:database]
+
+      rollback_migration(version, database)
+
+      redirect_to migrations_path, notice: "Migration #{version} has been rolled back."
+    end
+
+    private
+
+    def load_migrations
```
Comment (ka8725) [inline app/controllers/actual_db_schema/migrations_controller.rb:25]: Load migrations which helper method. Avoid mutating variables and using @ vars

**Review by ka8725 — CHANGES_REQUESTED** [review]

File: lib/actual_db_schema.rb [hunk 9]
```diff
@@ -30,6 +30,8 @@ class << self
     auto_rollback_disabled: ENV["ACTUAL_DB_SCHEMA_AUTO_ROLLBACK_DISABLED"].present?
   }
 
+  MigrationStruct = Struct.new(:status, :version, :name, :branch, :database, :filename)
```
Comment (ka8725) [inline lib/actual_db_schema.rb:33]: Use keyword init? https://ruby-doc.org/core-2.5.1/Struct.html#:~:text=If%20the%20optional%20keyword_init%20keyword%20argument%20is%20set%20to%20true%2C%20.new%20takes%20keyword%20arguments%20instead%20of%20normal%20arguments.

File: app/controllers/actual_db_schema/migrations_controller.rb (hunk 7 above)
Comment (VladislavSokov) [reply app/controllers/actual_db_schema/migrations_controller.rb]: Caching would cause incorrect behavior because for_each_db_connection runs actions for each database configuration. Caching data from one database might lead to using outdated or incorrect data from another database.

File: app/helpers/actual_db_schema/migrations_helper.rb [hunk 10]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module ActualDbSchema
+  # Helper methods for loading and displaying migrations.
+  module MigrationsHelper
+    def load_migrations
```
Comment (ka8725) [inline app/helpers/actual_db_schema/migrations_helper.rb:6]: By helper method I meant this https://stackoverflow.com/a/3993323

You can find a lot of examples using it inside BudgetingKid

**Review by ka8725 — CHANGES_REQUESTED** [review]

File: app/controllers/actual_db_schema/migrations_controller.rb (hunk 7 above)
Comment (ka8725) [reply app/controllers/actual_db_schema/migrations_controller.rb]: Every time a migration struct gets constructed (that happens on each phantom migration), it reads this CSV file from sctract. And this can be a performance issue. We should prevent it. Cache metadata in the controller for each connection?

File: lib/actual_db_schema.rb [hunk 11]
```diff
@@ -26,9 +27,12 @@ class << self
   self.failed = []
   self.config = {
     enabled: Rails.env.development?,
-    auto_rollback_disabled: ENV["ACTUAL_DB_SCHEMA_AUTO_ROLLBACK_DISABLED"].present?
+    auto_rollback_disabled: ENV["ACTUAL_DB_SCHEMA_AUTO_ROLLBACK_DISABLED"].present?,
+    ui_disabled: Rails.env.production? || ENV["ACTUAL_DB_SCHEMA_UI_DISABLED"].present?
```
Comment (ka8725) [inline lib/actual_db_schema.rb:31]:
```suggestion
    ui_enabled: Rails.env.development? || ENV["ACTUAL_DB_SCHEMA_UI_DISABLED"].present?
```

Let's revert this, I guess this thing can even be helpful for staging or other env. But we definitely know, that it's useful in one env and it's `development`. Only this should be the default one. Optionally, it can be enabled everywhere and env var is good for that. Moreover, `ui_enabled` corresponds with the convention above with `enabled`.

File: README.md [hunk 12]
```diff
@@ -62,6 +62,33 @@ The gem offers the following rake tasks that can be manually run according to yo
...
+```
+This page will display a list of phantom migrations for each database connection and provide options to view details and rollback migrations.
+
+## Disabling the UI
+
+By default, the UI is enabled. If you prefer to disable the UI, you can do so in two ways:
+
+### 1. Using Environment Variable
+
+Set the environment variable `ACTUAL_DB_SCHEMA_UI_DISABLED` to `true`:
```
Comment (ka8725) [inline README.md:79]: Fix with `ACTUAL_DB_SCHEMA_UI_ENABLED` env var.

File: README.md [hunk 13]
```diff
@@ -62,6 +62,33 @@ The gem offers the following rake tasks that can be manually run according to yo
 - `rails db:rollback_branches:manual` - run it to manually rolls back phantom migrations one by one.
 - `rails db:phantom_migrations` - displays a list of phantom migrations.
 
+## Accessing the UI
+
+The UI for managing migrations is enabled automatically. To access the UI, simply navigate to the following URL in your web browser:
+```
+http://localhost:3000/rails/migrations
+```
+This page will display a list of phantom migrations for each database connection and provide options to view details and rollback migrations.
```
Comment (ka8725) [inline README.md:71]:
```suggestion
This page displays a list of phantom migrations for each database connection and provides options to view details and rollback them.
```

File: README.md [hunk 14]
```diff
@@ -62,6 +62,33 @@ The gem offers the following rake tasks that can be manually run according to yo
...
+
+```sh
+export ACTUAL_DB_SCHEMA_UI_DISABLED=true
+```
+
+### 2. Using Initializer
+Add the following line to your initializer file (`config/initializers/actual_db_schema.rb`):
+
+```ruby
+ActualDbSchema.config[:ui_disabled] = true
```
Comment (ka8725) [inline README.md:89]:
```suggestion
ActualDbSchema.config[:ui_enabled] = true
```

File: README.md [hunk 15]
```diff
@@ -62,6 +62,33 @@ The gem offers the following rake tasks that can be manually run according to yo
...
+
+The UI for managing migrations is enabled automatically. To access the UI, simply navigate to the following URL in your web browser:
+```
+http://localhost:3000/rails/migrations
+```
+This page will display a list of phantom migrations for each database connection and provide options to view details and rollback migrations.
+
+## Disabling the UI
+
+By default, the UI is enabled. If you prefer to disable the UI, you can do so in two ways:
```
Comment (ka8725) [inline README.md:75]:
```suggestion
By default, the UI is enabled in the development environment. If you prefer to enable the UI for another environment, you can do so in two ways:
```

File: README.md (hunk 14 above)
Comment (ka8725) [reply README.md:89]: Add text:

```
With this option, the UI can be disabled for all environments or be enabled in specific ones.
```

File: lib/actual_db_schema/database_connection.rb [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+    def context
+      ar_version = Gem::Version.new(ActiveRecord::VERSION::STRING)
+      @context ||= if ar_version >= Gem::Version.new("7.2.0") ||
+                      (ar_version >= Gem::Version.new("7.1.0") && ar_version.prerelease?)
+                     ActiveRecord::Base.connection_pool.migration_context
+                   else
+                     ActiveRecord::Base.connection.migration_context
+                   end
+      @context.extend(ActualDbSchema::Patches::MigrationContext)
```
Comment (ka8725) [inline lib/actual_db_schema/database_connection.rb:34]: Are you sure this line is ok? How many times does it get executed?

**Review by ka8725 — CHANGES_REQUESTED** [review]

File: lib/actual_db_schema/commands/base.rb [hunk 17]
```diff
@@ -4,6 +4,10 @@ module ActualDbSchema
   module Commands
     # Base class for all commands
     class Base
+      def initialize(context: nil)
```
Comment (ka8725) [inline lib/actual_db_schema/commands/base.rb:7]:
```suggestion
      def initialize(context)
```

Why it's an optional param? Isn't it mandatory?

File: lib/actual_db_schema/commands/base.rb [hunk 18]
```diff
@@ -18,21 +22,7 @@ def call_impl
...
-      def fetch_migration_context
-        ar_version = Gem::Version.new(ActiveRecord::VERSION::STRING)
-        if ar_version >= Gem::Version.new("7.2.0") ||
-           (ar_version >= Gem::Version.new("7.1.0") && ar_version.prerelease?)
-          ActiveRecord::Base.connection_pool.migration_context
-        else
-          ActiveRecord::Base.connection.migration_context
-        end
-      end
+      attr_reader :context
```
Comment (ka8725) [inline lib/actual_db_schema/commands/base.rb:25]: Move up as usually done?

File: lib/actual_db_schema/commands/list.rb [hunk 19]
```diff
@@ -4,6 +4,10 @@ module ActualDbSchema
   module Commands
     # Shows the list of phantom migrations
     class List < Base
+      def initialize(context: nil)
```
Comment (ka8725) [inline lib/actual_db_schema/commands/list.rb:7]: Why it's needed? Isn't it working on base?

File: lib/actual_db_schema/commands/list.rb [hunk 20]
```diff
@@ -4,6 +4,10 @@ module ActualDbSchema
...
+        super(context: context)
+      end
```
Comment (ka8725) [inline lib/actual_db_schema/commands/list.rb:9]:
```suggestion
```

File: lib/actual_db_schema/commands/rollback.rb [hunk 21]
```diff
@@ -4,9 +4,9 @@ module ActualDbSchema
   module Commands
     # Rolls back all phantom migrations
     class Rollback < Base
-      def initialize(manual_mode: false)
+      def initialize(manual_mode: false, context: nil)
```
Comment (ka8725) [inline lib/actual_db_schema/commands/rollback.rb:7]:
```suggestion
      def initialize(context, manual_mode: false)
```

File: lib/actual_db_schema/commands/rollback.rb [hunk 22]
```diff
@@ -4,9 +4,9 @@ module ActualDbSchema
...
         @manual_mode = manual_mode || manual_mode_default?
-        super()
+        super(context: context)
```
Comment (ka8725) [inline lib/actual_db_schema/commands/rollback.rb:9]: Here that makes sense, unlike in others.

File: lib/actual_db_schema/database_connection.rb [hunk 23]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+module ActualDbSchema
+  # The class manages connections to each database and provides the appropriate migration context for each connection.
+  class DatabaseConnection
```
Comment (ka8725) [inline lib/actual_db_schema/database_connection.rb:5]:
```suggestion
  class MigrationContext
```

File: lib/actual_db_schema/database_connection.rb [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+    include Singleton
+
+    def for_each_migration_context
```
Comment (ka8725) [inline lib/actual_db_schema/database_connection.rb:8]:
```suggestion
    def each
```

File: lib/actual_db_schema.rb [hunk 25]
```diff
@@ -17,16 +20,15 @@
...
-
   class << self
     attr_accessor :config, :failed
   end
 
   self.failed = []
   self.config = {
-    enabled: Rails.env.development?,
-    auto_rollback_disabled: ENV["ACTUAL_DB_SCHEMA_AUTO_ROLLBACK_DISABLED"].present?
+    enabled: Rails.env.development? || ENV["ACTUAL_DB_SCHEMA_ENABLED"].present?,
```
Comment (ka8725) [inline lib/actual_db_schema.rb:29]: This option is not related to this PR, please extract it to separate one.

File: CHANGELOG.md [hunk 26]
```diff
@@ -1,3 +1,6 @@
+## [0.7.6] - 2024-07-02
+- Added UI
```
Comment (ka8725) [inline CHANGELOG.md:2]: mention env vars/configs added

File: app/controllers/actual_db_schema/migrations_controller.rb [hunk 27]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module ActualDbSchema
+  # Controller to display the list of phantom migrations for each database connection.
+  class MigrationsController < ActionController::Base
+    def index; end
+
+    def show
+      render :not_found, status: 404 unless migration
```
Comment (ka8725) [inline app/controllers/actual_db_schema/migrations_controller.rb:9]: What does it render now? Please check.

File: test/controllers/actual_db_schema/migrations_controller_test.rb [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+      Rails.logger = Logger.new($stdout)
+      ActionController::Base.view_paths = [File.expand_path("../../../app/views/", __dir__)]
+      active_record_setup
+      @utils.cleanup(TestingState.db_config)
+      @utils.prepare_phantom_migrations(TestingState.db_config)
+    end
+
+    def routes_setup
+      @routes = @app.routes
+      Rails.application.routes.draw do
```
Comment (ka8725) [inline test/controllers/actual_db_schema/migrations_controller_test.rb:21]: Still, why we should init routes here and why they are not initiated during engine loading? Maybe the problem is that the engine is not being loaded in tests? Then we should try to load it in tests?

These lines signify that the engine is not being loaded in tests and those lines are not covered by tests. And this is bad as the tests are fragile.

File: lib/actual_db_schema/patches/migration_context.rb [hunk 29]
```diff
@@ -69,21 +69,13 @@ def migrate(migration)
...
-                       end
+      def metadata
+        @metadata ||= ActualDbSchema::Store.instance.read
       end
 
       def branch_for(version)
         metadata.fetch(version, {})[:branch] || "unknown"
       end
-
-      def metadata
```
Comment (ka8725) [inline lib/actual_db_schema/patches/migration_context.rb:84]: Can you please move these lines back so that we don't have the not relevant diff in git?

**Comment by ka8725** [comment]
Manual testing results:

1. down migration cannot be rolled back, so hide the button?
image
2. If no phantom migrations the table looks clunky
image
Show some text saying no phantom migrations instead of the empty table?

**Comment by VladislavSokov** [comment]
> 3\. WDYT if we don't show down phantom migrations in the table at all? Reason - there is no point in running it up and also since it's down the user cannot do with it anything. This kind of migration is useless for the current branch too. So if it's hidden the user won't notice. It's not a relevant thing that can be destroyed. I tend to think we can remove it entirely on down. Please tell what do you think, thanks!

I think this is a great idea
